plugins/rh-rbac-backend/src/service/policy-builder.ts

```diff
@@ -68,27 +110,308 @@ export class PolicyBuilder {
+      // for ORMTypeAdapter.
+      // So, let's compensate this combination delete + create.
+      const isRemoved = await enforcer.removePolicy(...oldPolicy);
+      if (!isRemoved) {
+        throw new ServiceUnavailableError(); // 500
+      }
+
+      const isAdded = await enforcer.addPolicy(...newPolicy);
+      if (!isAdded) {
+        throw new ServiceUnavailableError(); // 500
+      }
+
+      resp.status(200).end();
     });
 
     return router;
   }
 }
+
+function getEntityReference(req: Request): string {
+  const str = req.params.namespace.concat('/');
+  return str + req.params.id;
+}
+
+function validatePolicyQueries(request: Request): Error | undefined {
+  if (!request.query.permission) {
+    return new Error('specify "permission" query param.');
+  }
+
+  if (!request.query.policy) {
+    return new Error('specify "policy" query param.');
+  }
+
+  if (!request.query.effect) {
+    return new Error('specify "effect" query param.');
+  }
+
+  return undefined;
+}
+
+function validatePolicy(policy: EntityReferencedPolicy): Error | undefined {
+  const err = validateEntityReference(policy.entityReference);
+  if (err) {
+    return err;
+  }
+
+  if (!policy.permission) {
+    return new Error(`'permission' field must not be empty`);
+  }
+
+  if (!policy.policy) {
+    return new Error(`'policy' field must not be empty`);
+  }
+
+  if (!policy.effect) {
+    return new Error(`'effect' field must not be empty`);
+  } else if (!isValidEffectValue(policy.effect)) {
+    return new Error(
+      `'effect' has invalid value: '${
+        policy.effect
+      }'. It should be: '${AuthorizeResult.ALLOW.toLocaleLowerCase()}' or '${AuthorizeResult.DENY.toLocaleLowerCase()}`,
+    );
+  }
+
+  return undefined;
+}
+
+function isValidEffectValue(effect: string): boolean {
+  return (
+    effect === AuthorizeResult.ALLOW.toLocaleLowerCase() ||
+    effect === AuthorizeResult.DENY.toLocaleLowerCase()
+  );
+}
+
+function validateEntityReference(entityRef?: string): Error | undefined {
+  if (!entityRef) {
+    return new Error(`'entityReference' must not be empty`);
+  }
+  try {
+    parseEntityRef(entityRef);
+  } catch (error) {
+    return error as Error;
+  }
+  return undefined;
+}
+
+function transformPolicyArray(
+  ...policies: string[][]
+): EntityReferencedPolicy[] {
+  return policies.map((p: string[]) => {
+    const [entityReference, permission, policy, effect] = p;
+    return { entityReference, permission, policy, effect };
+  });
+}
+
+function transformPolicyToArray(policy: EntityReferencedPolicy) {
+  return [
+    policy.entityReference!,
+    policy.permission!,
+    policy.policy!,
+    policy.effect!,
+  ];
+}
+
+function getFirstQuery(
+  queryValue: string | string[] | ParsedQs | ParsedQs[] | undefined,
+): string {
+  if (!queryValue) {
+    return '';
+  }
+  if (Array.isArray(queryValue)) {
+    if (typeof queryValue[0] === 'string') {
+      // return queryValue[0]
+      return queryValue.toString();
```

Comment:
Is this returning the correct value?

Reply:
Nice catch, It was correct only for single value query. I fixed it.

Reply:
Also I applied tests.